File: Archived2/Java/Silver/24445.java
import java.io.*;
import java.util.*;


public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static ArrayList<ArrayList<Integer>> adj;
    private static int order = 1;
    private static int startVertex;
    private static int[] visited;
    private static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        settingInput();
        bfs(startVertex);
        for (int i=1;i<visited.length;i++) sb.append(visited[i]).append("\n");
        System.out.println(sb);
        br.close();
    }

    private static void settingInput() throws IOException {
        int[] initData = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        visited = new int[initData[0]+1];
        startVertex = initData[2];
        adj = new ArrayList<>();
        for(int i=0;i<initData[0]+1; i++) adj.add(new ArrayList<>());

        ArrayList<int[]> edges = new ArrayList<>();
        for (int i=0;i<initData[1];i++) {
            int[] edge = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            edges.add(edge);
            edges.add(new int[] {edge[1], edge[0]});
        }

        edges.sort((a,b) -> {
            if (a[0] == b[0]) return b[1]-a[1];
            else return b[0]-a[0];
        });
        edges.forEach( (edge) -> adj.get(edge[0]).add(edge[1]));
    }

    private static void bfs(int vertex) {
        Queue<Integer> queue = new ArrayDeque<>();
        queue.add(vertex);
        visited[vertex] = order++;
        while (!queue.isEmpty()) {
            int itm = queue.poll();
            adj.get(itm).forEach( (next) -> {
                if (visited[next] == 0) {
                    queue.offer(next);
                    visited[next] = order++;
                }
            });
        }

    }

}

File: Archived2/Kotlin/Silver/24445.kt
import java.io.BufferedReader
import java.io.InputStreamReader

val adj = ArrayList<ArrayList<Int>>()
lateinit var visited: IntArray
var order = 1
var startVertex = 0


fun main() {

    settingInput()

    bfs(startVertex)

    println( buildString {
        for (i in 1 until visited.size) {
            append(visited[i]).append("\n")
        }
    } )
}

fun settingInput() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val initData = br.readLine().split(" ").map { it.toInt() }
    startVertex = initData[2]
    visited = IntArray(initData[0]+1)
    repeat(initData[0]+1) { adj.add(ArrayList()) }
    val edges = ArrayList<IntArray>()
    repeat(initData[1]) {
        val edge = br.readLine().split(" ").map { it.toInt() }
        edges.add(intArrayOf(edge[0], edge[1]))
        edges.add(intArrayOf(edge[1], edge[0]))
    }
    edges.sortWith( compareByDescending<IntArray> {it[0]}.thenByDescending { it[1] })
    edges.forEach { edge -> adj.get(edge[0]).add(edge[1])}
}

fun bfs(vertex: Int) {
    val queue = ArrayDeque<Int>()
    queue.addLast(vertex)
    visited[vertex] = order++
    while (queue.isNotEmpty()) {
        val itm = queue.removeFirst()
        adj.get(itm).forEach { next ->
            if(visited[next] == 0) {
                visited[next] = order++
                queue.addLast(next)
            }
        }
    }
}